### ISSUE TITLE
[STORM-1565] Multi-Lang Performance Improvements

### ISSUE THREAD

**Issue by jira-importer** [apache/storm#5400]
<p>1. add <em>org.apache.storm.multilang.MessagePackSerializer</em><br/>
2. change default "topology.multilang.serializer" to <em>MessagePackSerializer</em></p>

<p>According to <a href="http://msgpack.org/" class="external-link" target="_blank" rel="nofollow noopener">http://msgpack.org/</a> : It's like JSON, but fast and small.</p>
<blockquote>
<p>MessagePack is an efficient binary serialization format. It lets you exchange data among multiple languages like JSON. But it's faster and smaller. Small integers are encoded into a single byte, and typical short strings require only one extra byte in addition to the strings themselves.</p></blockquote>

---
<details><summary><i>Originally reported by <a title="Xin Wang" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=vesense">vesense</a>, imported from: <a href="https://issues.apache.org/jira/browse/STORM-1565" target="_blank">Multi-Lang Performance Improvements</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Xin Wang" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=vesense">vesense</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>resolution</b>: Won't Fix
<li><b>resolved</b>: 2023-10-23T16:20:45+01:00
<li><b>imported</b>: 2025-01-24
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user dan-blanchard commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1136#issuecomment-191364958" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1136#issuecomment-191364958</a></p>

<p>    You may want to look at the <span class="error">&#91;Pyleus project&#39;s `MessagePackSerializer`&#93;</span>(<a href="https://github.com/Yelp/pyleus/blob/64036a8d8ccaa656ddebb3dbd1a7dfb40e15a14a/topology_builder/src/main/java/com/yelp/pyleus/serializer/MessagePackSerializer.java" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/Yelp/pyleus/blob/64036a8d8ccaa656ddebb3dbd1a7dfb40e15a14a/topology_builder/src/main/java/com/yelp/pyleus/serializer/MessagePackSerializer.java</a>), because I know they had to add some things in to handle weird edge cases.  I'm not really familiar enough with implementing custom serializers to recognize why your approach and theirs look rather different.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user dan-blanchard commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1136#discussion_r54769764" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1136#discussion_r54769764</a></p>

<p>    &#8212; Diff: conf/defaults.yaml &#8212;<br/>
    @@ -226,7 +226,7 @@ topology.eventlogger.executors: null<br/>
     topology.tasks: null</p>
<ol>
	<li>maximum amount of time a message has to complete before it's considered failed<br/>
     topology.message.timeout.secs: 30<br/>
    -topology.multilang.serializer: "org.apache.storm.multilang.JsonSerializer"<br/>
    +topology.multilang.serializer: "org.apache.storm.multilang.MessagePackSerializer"
	<ul class="alternate" type="square">
		<li>
		<ul class="alternate" type="square">
			<li>End diff &#8211;</li>
		</ul>
		</li>
	</ul>
	</li>
</ol>


<p>    While having the `MessagePackSeriailzer` built-in would be amazing, changing it to the default serializer would break a lot of multi-lang libraries, so I don't think that's a great idea.  Let people opt in.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user vesense commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1136#issuecomment-193103574" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1136#issuecomment-193103574</a></p>

<p>    Thanks @dan-blanchard I'd love to hear from pyleus contributors.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user dan-blanchard commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1136#discussion_r60652163" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1136#discussion_r60652163</a></p>

<p>    &#8212; Diff: pom.xml &#8212;<br/>
    @@ -202,6 +202,7 @@<br/>
     <clj-time.version>0.8.0</clj-time.version><br/>
     <curator.version>2.9.0</curator.version><br/>
     <json-simple.version>1.1</json-simple.version><br/>
    +<msgpack.version>0.6.12</msgpack.version><br/>
    &#8212; End diff &#8211;</p>

<p>    This would be <b>much</b> more useful if it were implemented using a newer version of the `msgpack-core` library—they changed the name after 0.6.12—because 0.7 and above supports the `BINARY` format, which lets you send arbitrary bytes.  Without that, you won't be able to send tuples containing arbitrary bytes with this serializer.  This is also a problem with the JSON serializer (because JSON strings can't contain non-Unicode characters), but it would be great if we didn't have the problem here.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user dan-blanchard commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1136#discussion_r60652344" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1136#discussion_r60652344</a></p>

<p>    &#8212; Diff: pom.xml &#8212;<br/>
    @@ -202,6 +202,7 @@<br/>
     <clj-time.version>0.8.0</clj-time.version><br/>
     <curator.version>2.9.0</curator.version><br/>
     <json-simple.version>1.1</json-simple.version><br/>
    +<msgpack.version>0.6.12</msgpack.version><br/>
    &#8212; End diff &#8211;</p>

<p>    Annoyingly, they changed the whole API with the change from `msgpack` to `msgpack-core`, so the template approach won't work anymore.  You can see new usage examples <span class="error">&#91;here&#93;</span>(<a href="https://github.com/msgpack/msgpack-java/blob/develop/msgpack-core/src/test/java/org/msgpack/core/example/MessagePackExample.java" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/msgpack/msgpack-java/blob/develop/msgpack-core/src/test/java/org/msgpack/core/example/MessagePackExample.java</a>).</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user vesense commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1136#discussion_r60681547" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1136#discussion_r60681547</a></p>

<p>    &#8212; Diff: pom.xml &#8212;<br/>
    @@ -202,6 +202,7 @@<br/>
     <clj-time.version>0.8.0</clj-time.version><br/>
     <curator.version>2.9.0</curator.version><br/>
     <json-simple.version>1.1</json-simple.version><br/>
    +<msgpack.version>0.6.12</msgpack.version><br/>
    &#8212; End diff &#8211;</p>

<p>    @dan-blanchard Thanks. I will update.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1136" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1136</a></p>

<p>    @vesense <br/>
    Sorry to reach out too late. Some comments for me:</p>

<p>    1. We might want to include this to multilang (like multilang-serializer) module so that it doesn't affect dependency if users don't use multilang feature. But I'm also OK to have this to core if we are OK to just shade msgpack.<br/>
    2. As @dan-blanchard stated we may want to have stable implementation of serializer if possible.<br/>
    3. I totally agree with @dan-blanchard that changing default behavior should be avoided unless releasing major version.<br/>
    4. It requires users to implement custom multi-lang logic to apply deserialization. IMO, I think it makes issue more complicated. We could provide implementation by default, but it will force user to have dependency at all, and more bad thing is that we need to address this at least three languages.</p>

<p>    Personally I feel that we're stuck on having less-maintaining and non-library-dependent default implementations. I can take care of python / ruby implementation though I'm not expert on these, but have no idea on nodejs.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1136" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1136</a></p>

<p>    So it may need to discuss with current state of multi-lang support once rather than adding something continuously.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user vesense commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1136" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1136</a></p>

<p>    @HeartSaVioR Agree with you. We should have a discussion about multi-lang support. (Can open a thread on dev@ mailing list)</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user dan-blanchard commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1136" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1136</a></p>

<p>    I'm not on the Storm dev mailing list yet—I guess I should probably fix that—but we actually already have a Python implementation of msgpack serialization in <span class="error">&#91;pystorm&#93;</span>(<a href="https://github.com/pystorm/pystorm" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/pystorm/pystorm</a>), the Python multi-lang implementation that powers <span class="error">&#91;streamparse&#93;</span>(<a href="https://github.com/Parsely/streamparse" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/Parsely/streamparse</a>).</p>

<p>    I've been meaning to propose for a while that instead of having your own default Python multi-lang implementation in Storm that very few people use (because it's not very Pythonic or production-ready), you should instead point people to use <span class="error">&#91;pystorm&#93;</span>(<a href="https://github.com/pystorm/pystorm" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/pystorm/pystorm</a>) at the very least. It provides all the functionality that the `storm.py` provides but with documentation, exception handling, logging, etc.  pystorm is intended to be used as a starting point for more full-fledged Python-Storm interop libraries (like streamparse), but if you want to replace the bare bones provdided by `storm.py` with an enhanced version of the same thing, it seems like pystorm is the way to go.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1136" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1136</a></p>

<p>    @dan-blanchard <br/>
    Yeah, agreed. I've addressed several multi-lang issues so I had been interested on multi-lang feature of Storm, but few people have interest on maintaining / improving basic implementation of multilang protocol in various language even python.<br/>
    Pystorm seems great, and we lack contributors who care multi-lang implementations as one of major module so I'd still rather rely on stable implementation than build from scratch. Just one concern for me is that it seems to be heavily depending on you (I mean individual) so I'm curious we feel safe to rely on.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1136" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1136</a></p>

<p>    @dan-blanchard And please note that opinions are my own, so we should raise this to discussion and let community give various opinions if we would really want to.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user dan-blanchard commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1136" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1136</a></p>

<p>    > Pystorm seems great, and we lack contributors who care multi-lang implementations as one of major module so I'd still rather rely on stable implementation than build from scratch. Just one concern for me is that it seems to be heavily depending on you (I mean individual) so I'm curious we feel safe to rely on.</p>

<p>    I work full-time for a company that uses pystorm and streamparse in production, and we have a team dedicated to maintaining these projects, so I don't think you need to worry about that much.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1136" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1136</a></p>

<p>    OK great. Could you subscribe dev@ mailing list? I occasionally initiate discussion from there (multilang, too) so you might want to have a talk regarding to multilang.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user dan-blanchard commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1136" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1136</a></p>

<p>    Yup. I subscribed yesterday.<br/>
    On Wed, Jun 22, 2016 at 8:35 PM Jungtaek Lim <notifications@github.com><br/>
    wrote:</p>

<p>    > OK great. Could you subscribe dev@ mailing list? I occasionally initiate<br/>
